rcl/src/rcl/arguments.c

```diff
@@ -354,6 +354,45 @@ rcl_get_zero_initialized_arguments(void)
   return default_arguments;
 }
 
+rcl_ret_t
+rcl_remove_ros_arguments(
+  char const * const argv[],
+  rcl_arguments_t * args,
+  rcl_allocator_t allocator,
+  int * nonros_argc,
+  const char ** nonros_argv[])
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(argv, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCL_CHECK_ARGUMENT_FOR_NULL(nonros_argc, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCL_CHECK_ARGUMENT_FOR_NULL(args, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  *nonros_argc = rcl_arguments_get_count_unparsed(args);
+  if (*nonros_argc > 0) {
+    int * unparsed_indices = NULL;
+    rcl_ret_t ret;
+    ret = rcl_arguments_get_unparsed(args, allocator, &unparsed_indices);
+
+    if (RCL_RET_OK != ret) {
+      allocator.deallocate(unparsed_indices, allocator.state);
```

Comment:
My assumption was that if `rcl_arguments_get_unparsed()` fails it shouldn't store anything in `unparsed_indices`. But that might not be the case. I think this check is safe and defensive, since you initialized it to null.

The only corner case that could be dangerous is that if it does set `unparsed_indices`, at the very least that function should have already free'd the memory it allocated for that spot. So I think this can result in a double free.

I'd consult with @sloretz and see what the error post condition of this function is, and my expectation is that this deallocate when it fails should not happen.

Reply:
I reviewed the implementation of `rcl_arguments_get_unparsed` and you are correct.  It either passes with `unparsed_indices` set to non-NULL or it fails and `unparsed_indices` is NULL.

Reply:
The only error cases in `rcl_arguments_get_unparsed` happen either when one of the arguments is invalid or when allocating the output fails, do deallocating here isn't needed. I would also expect any new error cases in `rcl_arguments_get_unparsed` to cause it to clean up after itself before returning.

I can update `rcl_arguments_get_unparsed` documentation to say deallocating is only necessary if it returns `RCL_RET_OK`.